### PR TITLE
now-cli 7.0.2 (new formula)

### PR DESCRIPTION
**Pull request by amio**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

For `brew install --build-from-source <formula>` requirement, this is a prebuilt binary, it should works on all recently versions macOS.

### CODE REVIEW

**Comment by leo** [comment]
Wonderful! Can you make the word "Realtime" in the description small and maybe call the class "NowCLI" instead of "NowCli"?

**Comment by bfontaine** [comment]
> maybe call the class "NowCLI" instead of "NowCli"?

The class name reflects the filename so it should be `NowCli`. `NowCLI` would be appropriate for a file named `now-c-l-i.rb`.

File: Formula/now-cli.rb
Comment (bfontaine) [inline Formula/now-cli.rb]: Use `bin.install "now-macos" => "now"`.

**Review by bfontaine — CHANGES_REQUESTED** [review]

**Comment by amio** [comment]
@leo Yeah, sure!
I've fixed the uppercase mistake, while homebrew have a class [naming convention](http://docs.brew.sh/Formula-Cookbook.html#a-quick-word-on-naming), so I just leave it `NowCli` right now. Since the class name is using internal, I think it would be just fine. :P

File: Formula/now-cli.rb
Comment (amio) [inline Formula/now-cli.rb]: Thanks for point out!

File: Formula/now-cli.rb
Comment (ilovezfs) [inline Formula/now-cli.rb]: We'll need a test that exercises and verifies the software's functionality here.

**Review by ilovezfs — CHANGES_REQUESTED** [review]

File: Formula/now-cli.rb
Comment (ilovezfs) [inline Formula/now-cli.rb]: We'll need to build the software from source.

**Review by ilovezfs — CHANGES_REQUESTED** [review]

File: Formula/now-cli.rb
Comment (amio) [inline Formula/now-cli.rb]: I borrowed this simple test from an other package, thought a full feature tests would be overwhelming, then usually how many feature should I test?

File: Formula/now-cli.rb
Comment (amio) [inline Formula/now-cli.rb]: I thought [`bottle :unneeded`](http://www.rubydoc.info/github/Homebrew/brew/Formula.bottle) is for prebuilt package, did I get it wrong?

And the now-cli was originally on npm registry, which mainly download a prebuilt binary while installation, the team at Zeit also create [zeit/pkg](https://github.com/zeit/pkg) for this, for "No need to download hundreds of files via npm install to deploy your application. Deploy it as a single file"(one of [Use Cases](https://github.com/zeit/pkg#use-cases)), I think it was meant to be installed as a prebuilt binary.

ping @leo Just in case I have any misunderstanding :D

**Comment by leo** [comment]
That's correct. Now CLI doesn't cannot be expected to work properly anymore went built from source (since our code expects the node version that we ship with the binary - it contains node). No need to build from source!

**Comment by ilovezfs** [comment]
There are other formulae in core with run-time node dependencies (e.g. kibana). In order to be included in Homebrew, it is a requirement that the software be built from source.

File: Formula/now-cli.rb
Comment (ilovezfs) [inline Formula/now-cli.rb]: Take a look at `fsql` and `govendor` for some test examples.

Comment (amio) [inline Formula/now-cli.rb]: Thanks!

**Comment by leo** [comment]
Are you saying it needs to be built from source on install? Why that?

**Comment by ilovezfs** [comment]
@leo the users end up pouring our "bottles" (the binaries we build from source), unless they explicitly request their own source build. But the bottles themselves are built by our CI from source as a matter of policy. Anything not built from source by us would need to be distributed via Homebrew Cask, but most CLI tools are ineligible for Homebrew Cask as we like to have CLI tools in homebrew/core (this repository), and we also like to avoid duplication with casks.

**Comment by leo** [comment]
@ilovezfs I see. We already have "now" on cask, but that's our Desktop app. So now we'd like to distribute our CLI over homebrew as well. Can you tell me and @amio how we should tell homebrew to build our CLI? Like... the config?

**Comment by ilovezfs** [comment]
@leo how do you usually build the CLI?

**Comment by ilovezfs** [comment]
@leo usually we do this for node formulae:
```
  def install
    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
    bin.install_symlink Dir["#{libexec}/bin/*"]
  end
```

However, I'm getting the following error:
```
download	license.md	package.json	scripts
bash-3.2$ npm pack
npm WARN prepublish-on-install As of npm@5, `prepublish` scripts are deprecated.
npm WARN prepublish-on-install Use `prepare` for build steps and `prepublishOnly` for upload-only.
npm WARN prepublish-on-install See the deprecation note in `npm help scripts` for more information.

> now@7.0.2 prepublish /private/tmp/now-cli-20170624-92858-z4twph/now-cli-7.0.2
> in-install || (npm run webpack && cp /dev/null download/dist/now)

sh: in-install: command not found

> now@7.0.2 webpack /private/tmp/now-cli-20170624-92858-z4twph/now-cli-7.0.2
> webpack --context download --config download/webpack.js

sh: webpack: command not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! now@7.0.2 webpack: `webpack --context download --config download/webpack.js`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the now@7.0.2 webpack script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/joe/Library/Caches/Homebrew/npm_cache/_logs/2017-06-24T16_40_33_483Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! now@7.0.2 prepublish: `in-install || (npm run webpack && cp /dev/null download/dist/now)`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the now@7.0.2 prepublish script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/joe/Library/Caches/Homebrew/npm_cache/_logs/2017-06-24T16_40_33_507Z-debug.log
```

Have you got this working with npm 5 yet?

**Comment by leo** [comment]
@ilovezfs No, I personally haven't tried npm 5 yet. But it should work just fine! @amio are you interested in tackling this? If not, I'll do it.

**Comment by ilovezfs** [comment]
@leo in case it helps, here's what the node DSL is doing behind the scenes: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/language/node.rb

**Comment by ilovezfs** [comment]
@leo if I add `depends_on "webpack" => :build` to the formula, then the error changes to
```
bash-3.2$ npm pack
npm WARN prepublish-on-install As of npm@5, `prepublish` scripts are deprecated.
npm WARN prepublish-on-install Use `prepare` for build steps and `prepublishOnly` for upload-only.
npm WARN prepublish-on-install See the deprecation note in `npm help scripts` for more information.

> now@7.0.2 prepublish /private/tmp/now-cli-20170624-94450-wyouwm/now-cli-7.0.2
> in-install || (npm run webpack && cp /dev/null download/dist/now)

sh: in-install: command not found

> now@7.0.2 webpack /private/tmp/now-cli-20170624-94450-wyouwm/now-cli-7.0.2
> webpack --context download --config download/webpack.js

Hash: 1fa89461ea43baf4bc5c
Version: webpack 3.0.0
Time: 37ms
      Asset     Size  Chunks             Chunk Names
download.js  2.59 kB       0  [emitted]  main
   [0] multi ./src/index.js 28 bytes {0} [built]

ERROR in multi ./src/index.js
Module not found: Error: Can't resolve 'babel-loader' in '/private/tmp/now-cli-20170624-94450-wyouwm/now-cli-7.0.2/download'
 @ multi ./src/index.js
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! now@7.0.2 webpack: `webpack --context download --config download/webpack.js`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the now@7.0.2 webpack script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/joe/Library/Caches/Homebrew/npm_cache/_logs/2017-06-24T16_58_40_739Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! now@7.0.2 prepublish: `in-install || (npm run webpack && cp /dev/null download/dist/now)`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the now@7.0.2 prepublish script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/joe/Library/Caches/Homebrew/npm_cache/_logs/2017-06-24T16_58_40_760Z-debug.log
bash-3.2$
```

**Comment by leo** [comment]
Yea there's something really wrong with how homebrew builds stuff. Each of the dependencies used while building have a lot of sub dependencies, so I don't understand how it makes sense to define each of them... 🤔

**Comment by leo** [comment]
Also wouldn't be possible, really. Since the dependencies change all the time.

**Comment by ilovezfs** [comment]
These are all using the node DSL successfully:

alexjs.rb:1:require "language/node"
angular-cli.rb:1:require "language/node"
autocode.rb:1:require "language/node"
azure-cli.rb:1:require "language/node"
babel.rb:1:require "language/node"
bit.rb:1:require "language/node"
bower.rb:1:require "language/node"
chronograf.rb:1:require "language/node"
coffeescript.rb:1:require "language/node"
diff-so-fancy.rb:1:require "language/node"
elixirscript.rb:1:require "language/node"
firebase-cli.rb:1:require "language/node"
generate-json-schema.rb:1:require "language/node"
gitter-cli.rb:1:require "language/node"
grunt-cli.rb:1:require "language/node"
ios-sim.rb:1:require "language/node"
jhipster.rb:1:require "language/node"
jsdoc3.rb:1:require "language/node"
jsonlint.rb:1:require "language/node"
kibana.rb:1:require "language/node"
kibana@4.1.rb:1:require "language/node"
kibana@4.4.rb:1:require "language/node"
nativefier.rb:1:require "language/node"
svgo.rb:1:require "language/node"
typescript.rb:1:require "language/node"
webpack.rb:1:require "language/node"
write-good.rb:1:require "language/node"

So I'm sure this can be figured out :)

**Comment by ilovezfs** [comment]
@leo yeah this seems to be an issue with running `npm pack`. If I use our pre-npm-5 node DSL, it seems to work fine.

**Comment by leo** [comment]
@ilovezfs Can we select a node version to build on?

**Comment by ilovezfs** [comment]
```
    system "npm", "install", "babel-loader", "babel-core",
           "babel-plugin-transform-async-to-generator",
           "babel-plugin-transform-runtime", "babel-preset-es2015", "death",
           "node-fetch", "progress", *Language::Node.local_npm_install_args
    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
    bin.install_symlink Dir["#{libexec}/bin/*"]
```
is what works with the pre-npm-5 DSL.

>Can we select a node version to build on?

You could try to use `node@6`, but it will still have npm 5.

**Comment by ilovezfs** [comment]
@leo OK, so cheating works:
```
  def install
    args = Language::Node.local_npm_install_args << "npm@4.0.0"
    system "npm", "install", "--verbose", *args
    npm = "node_modules/npm/bin/npm-cli.js"
    system npm, "install", "babel-loader", "babel-core",
                "babel-plugin-transform-async-to-generator",
                "babel-plugin-transform-runtime", "babel-preset-es2015", "death",
                "node-fetch", "progress", *Language::Node.local_npm_install_args
    system npm, "install", "--verbose", "--global", "--prefix=#{libexec}"
    bin.install_symlink Dir["#{libexec}/bin/*"]
  end
```
Is there somewhere where babel, etc. are specified?

**Comment by leo** [comment]
[Yep!](https://github.com/zeit/now-cli/blob/master/package.json#L72-L75) But they're not all in one separate config prop (they're usually devDeps), due to our build system.

**Comment by ilovezfs** [comment]
OK, so that would give us
```
  def install
    args = Language::Node.local_npm_install_args << "npm@4.0.0"
    system "npm", "install", "--verbose", *args
    npm = "node_modules/npm/bin/npm-cli.js"
    system npm, "install", "babel-core@6.25.0", "babel-loader@7.0.0",
                           "babel-plugin-transform-async-to-generator@6.24.1",
                           "babel-plugin-transform-runtime@6.23.0",
                           "babel-preset-es2015@6.24.1", "death@1.1.0",
                           "node-fetch@1.7.1", "progress@2.0.0"
    system npm, "install", "--verbose", "--global", "--prefix=#{libexec}", "."
    bin.install_symlink Dir["#{libexec}/bin/*"]
  end
```
for npm 4.

**Comment by ilovezfs** [comment]
@chrmoritz any thoughts here on how to get this working sanely with the node DSL (i.e., with dependencies not having to be manually specified, and the npm 5 pack-and-install-tarball routine not dying)?

**Comment by chrmoritz** [comment]
Unfortunaly this is a bug in our new `language node`, which does not support running dependencies in a prepublish script yet. 

This is because we use `npm pack` inside `language node` atm only as a way to pack the package without it's depenencies to make it installable with `npm install` (without the symlink issues from npm@5 we had) and are relying on `npm install` to install all the dependencies for us. But this obviously breaks the case, if one of the dependencies (in this case webpack) is required to pack the applicatation itself.

A litttle background:
* npm script adds the `bin` of all it's depenendencies to the users `PATH` before executing npm script the simplify writing them
* npm (every version) calls `prepublish` before (at the beginning of) every `npm pack`
* npm@<=4 even calls `prepublish` before every `npm install` (but our old `language node` wasn't afffected because we had all dependcies in place at npm install time
* but with our new language node we have no dependencies in place at npm pack time

Working on a fix for this now. Unfortunaly this PR is blocked until the fix for `language node` gets merged.

**Comment by leo** [comment]
I think there's some confusion in the room right now. We're using `scripts.pack` for building our software, so we need to run `npm run pack` to run it. Seems like it has the same name as your script for bundling the code but it's actually doing something different?

**Comment by chrmoritz** [comment]
@leo Sry, for confusing you. My last comment was primary directed to @ilovezfs and therefore expected some knowledge about `language/node`'s internals.

However you will be able to build `now-cli` successfully from source after https://github.com/Homebrew/brew/pull/2820 gets merged with this formula:

```ruby
require "language/node"

class NowCli < Formula
  desc "The command-line interface for realtime global deployments."
  homepage "https://zeit.co/now"
  url "https://github.com/zeit/now-cli/archive/7.0.2.tar.gz"
  sha256 "7924feb98f07a599cb9ffdabd07f95048864cf4fe15fdcbd2922b8cd8b32a8be"

  depends_on "node"

  def install
    system "npm", "install", *Language::Node.std_npm_install_args(libexec, prepublishRequiresDeps=true)
    bin.install_symlink Dir["#{libexec}/bin/*"]
  end

  test do
    system "#{bin}/now", "-v"
  end
end
```

**Comment by amio** [comment]
Thanks for your help! @chrmoritz 

@leo For the testing issue, I see most features of now-cli requires user logged in to perform, except `now -v` and `now help`, is there some other things we could test without logging in?

**Comment by ilovezfs** [comment]
>BTW, I looked heroku which has similar features as now-cli, it also just verify the version command in tests.

@amio we're requiring more rigorous `test do` blocks for all new formulae, even though there are still some older formulae with "trivial" `test do` blocks  or no `test do` block at all.

**Comment by ilovezfs** [comment]
>Through this will still create a giant binary because it bundles a copy of node itself in it, which might to be desired for a homebrew formula.

@chrmoritz can we get it to use the `node` formula instead?

**Comment by amio** [comment]
> we're requiring more rigorous `test do` blocks for all new formulae

@ilovezfs totally understand. While for now-cli, a deployment tool for now.sh, almost every feature would require user logged in to perform, except `now -v` and `now -h`. Would it be OK to test those two command? Should I add some more to tests?

<details>
  <summary>Output of `now help`</summary>

    𝚫 now [options] <command | path>

    Commands:

      Cloud

        deploy               [path]      Performs a deployment (default)
        ls | list            [app]       List deployments
        rm | remove          [id]        Remove a deployment
        ln | alias           [id] [url]  Configures aliases for deployments
        domains              [name]      Manages your domain names
        certs                [cmd]       Manages your SSL certificates
        secrets              [name]      Manages your secret environment variables
        dns                  [name]      Manages your DNS records
        logs                 [url]       Displays the logs for a deployment
        scale                [args]      Scales the instance count of a deployment
        help                 [cmd]       Displays complete help for [cmd]

      Administrative

        billing | cc         [cmd]       Manages your credit cards and billing methods
        upgrade | downgrade  [plan]      Upgrades or downgrades your plan
        teams                [team]      Manages your teams
        switch                           Switches between teams and your account
        login                            Login into your account or creates a new one
        logout                           Logout from your account

    Options:

      -h, --help                Output usage information
      -v, --version             Output the version number
      -n, --name                Set the name of the deployment
      -c FILE, --config=FILE    Config file
      -d, --debug               Debug mode [off]
      -f, --force               Force a new deployment even if nothing has changed
      -t TOKEN, --token=TOKEN   Login token
      -L, --login               Configure login
      -l, --links               Copy symlinks without resolving their target
      -p, --public              Deployment is public (`/_src` is exposed) [on for oss, off for premium]
      -e, --env                 Include an env var (e.g.: `-e KEY=value`). Can appear many times.
      -E FILE, --dotenv=FILE    Include env vars from .env file. Defaults to '.env'
      -C, --no-clipboard        Do not attempt to copy URL to clipboard
      -N, --forward-npm         Forward login information to install private npm modules
      --session-affinity        Session affinity, `ip` (default) or `random` to control session affinity.

    Enforcable Types (when both package.json and Dockerfile exist):

      --npm                     Node.js application
      --docker                  Docker container
      --static                  Static file hosting

    Examples:

    – Deploys the current directory

      $ now

    – Deploys a custom path `/usr/src/project`

      $ now /usr/src/project

    – Deploys a GitHub repository

      $ now user/repo#ref

    – Deploys a GitHub, GitLab or Bitbucket repo using its URL

      $ now https://gitlab.com/user/repo

    – Deploys with ENV vars

      $ now -e NODE_ENV=production -e MYSQL_PASSWORD=@mysql-password

    – Displays comprehensive help for the subcommand `list`

      $ now help list
</details>

**Comment by cglong** [comment]
> almost every feature would require user logged in to perform

@amio Could you do something similar to [`firebase-cli`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/firebase-cli.rb), which automates the login flow and validates that the application requested a login?

**Comment by leo** [comment]
Just to avoid confusion here: It's very important that we're bundling node with the CLI, because we always depend on our own selection of the node instance we use.

**Comment by ilovezfs** [comment]
@leo we cannot bundle in a pre-built binary version of node.

**Comment by leo** [comment]
@ilovezfs No. I think there's still some confusion in here. The `npm run pack` command runs a locally installed dependency called [pkg](https://github.com/zeit/pkg), which bundles in node with the code (it's our own creating and it should work on your CI just fine).

**Comment by ilovezfs** [comment]
>The npm run pack command runs a locally installed dependency called pkg, which bundles in node with the code (it's our own creating and it should work on your CI just fine).

If this is a prebuilt binary node, which it sounds like it is, we'll need to have `now` use the Homebrew `node` formula instead.

**Comment by chrmoritz** [comment]
I was able to build this to run against a homebrew version of node at runtime (not bundled into the executable) with some tiny inreplaces:

```ruby
require "language/node"

class NowCli < Formula
  desc "The command-line interface for realtime global deployments."
  homepage "https://zeit.co/now"
  url "https://github.com/zeit/now-cli/archive/7.0.2.tar.gz"
  sha256 "7924feb98f07a599cb9ffdabd07f95048864cf4fe15fdcbd2922b8cd8b32a8be"

  depends_on "node"

  def install
    inreplace "package.json" do |s|
      s.gsub! '"now": "download/dist/now"', '"now": "bin/now.js"' # patch bin
      s.gsub! '"download/dist"', '"bin"' # patch files array
      s.gsub! '"download/install.js"', '"lib"' # patch files array
      s.gsub! '"prepublish"', '"no_prepublish_for_you"' # don't ever run prepublish
      s.gsub! '"postinstall"', '"no_postinstall_for_you"' # don't ever run "postinstall"
      s.gsub! '"devDependencies"', '"dependencies"' # some devDeps are real deps, ToDo: filter out real devDeps
    end
    inreplace "bin/now.js", "if (process.pkg) {", "if (process.pkg && false) {" # disable auto updater
    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
    bin.install_symlink Dir["#{libexec}/bin/*"]
  end

  test do
    system "#{bin}/now", "-v"
  end
end
```

I can confirm that help is outputting the right thing with it, through this will need more real testing.

**Comment by ilovezfs** [comment]
Is there a way to tell npm which package.json to use?

**Comment by chrmoritz** [comment]
> Is there a way to tell npm which package.json to use?

I guess no, but could you elaborate in more detail what you want to do? 
I do agree, that this huge inreplace isn't ideal, expecial because it handels seperating the upstream devDeps into real deps and real devDeps pretty well (see ToDo comment above). But the only way to fix this would be to either maintain our own package.json (which build now with a runtime dependency on homebrew node) or get this use case supported by upstream.

**Comment by ilovezfs** [comment]
So we're all on the same page, here is what the inreplaces look like as a patch:
https://gist.githubusercontent.com/ilovezfs/6ac566e1b2e30ba36cfcbd59ef3b65c0/raw/ff1160912b2096b8ac6fe176ed36531277baad94/gistfile1.txt

>I guess no, but could you elaborate in more detail what you want to do?

I'm wondering if upstream could provide a `package-homebrew.json` (or some other name) and then we could point npm at that, which would be preferable to say `mv package-homebrew.json package.json`, though that could work, too, I suppose.

**Comment by chrmoritz** [comment]
> So we're all on the same page, here is what the inreplaces look like as a patch:
https://gist.githubusercontent.com/ilovezfs/6ac566e1b2e30ba36cfcbd59ef3b65c0/raw/ff1160912b2096b8ac6fe176ed36531277baad94/gistfile1.txt

Through the patch could just remove the whole line with the `postinstall` and `prepublish` script and in the ideal world would not only rename devDeps to dependencies, but would also remove all devDeps from the list which are not needed at runtime, like webpack and babel.

> I'm wondering if upstream could provide a package-homebrew.json (or some other name) and then we could point npm at that, which would be preferable to say mv package-homebrew.json package.json, though that could work, too, I suppose.

We would have to do the `mv  package-homebrew.json package.json`, but yes getting this maintained by upstream would be the ideal solution.

**Comment by ilovezfs** [comment]
Yeah I think an install method that looked like this
```
  def install
    mv "package-homebrew.json", "package.json"
    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
    bin.install_symlink Dir["#{libexec}/bin/*"]
  end
```
would definitely be nice.

**Comment by ilovezfs** [comment]
I guess we'd still need to figure out what to do about the change to "bin/now.js" or is that something that the package.json could handle too?

**Comment by chrmoritz** [comment]
Yeah, through I'm still not 100% confident that the v8 version in node8 supports all the features needed to skip the babel transpilation step completly (atleast it supports `async` ). I was only able to test the help command wich works fine, but I have no idea if the commands wich requires a login will work too. 

Also we may want to package the same node version as upstream (similar to the kibana formulas). Because in the current approach we would use a different node version than upstream which on the one hand could lead to unexpected issues, but on the other hand would solve the 'bundled outdated node version with open security vulnerabilities'-madness we have in the kibana formulas right now.

**Comment by leo** [comment]
The transpilation step is for the script that gets uploaded to npm (which downloads the bundled binary and places it into npm's global package directory). There's no way to avoid transpiling this, since it needs to work on very old nodes as well (that's the reason why the main code comes with node, because it needs to run even on devices with very old node).

**Comment by ilovezfs** [comment]
The accommodation for kibana is not a pattern we want to see copied around. Ideally if a different node version than the main node formula must be used, then we'd want to use one of the Homebrew versioned @ formulae (node@6 or node@4) and set the shebangs accordingly, or use wrapper scripts. But using the main node formula is the best approach whenever feasible.

**Comment by ilovezfs** [comment]
@leo are you able to test @chrmoritz's formula from https://github.com/Homebrew/homebrew-core/pull/14827#issuecomment-310897963, which uses Homebrew node 8, to see if it works as expected? It builds for me locally.

**Comment by chrmoritz** [comment]
@leo If this gets packed into homebrew-core, homebrew will ensure that it is always run with the latest node version. Handling very old node version might be a use case for your provided binaries, but it will definitely not be important for the homebrew package. Also we try to avoid bundling any version of node into the now executable for the homebrew formula.

The current proposed approve does the following: 
* download the source tarball from github
* installs the bin and lib folder as it is with npm
* installs all dependencies listed under devDependencies (through we might want to filter out deps no needed at runtume like webpack and babel)
* run the bin/now.js script with a homebrew provided version of node 8

**Comment by chrmoritz** [comment]
@ilovezfs Or here is a version with a more complete patch: https://gist.github.com/chrmoritz/0073afc406fc36f2b1fc46f08352d8a9

```
brew install https://gist.githubusercontent.com/chrmoritz/0073afc406fc36f2b1fc46f08352d8a9/raw/31a663841b8647cf4380aeeb605be37ee0fc8408/now-cli.rb
```

**Comment by chrmoritz** [comment]
I think I have an even better alternative, which would both reduce the maintenance overhead on homebrew's side *(not having to maintain a giant inreplace or a patch every version bump)* and on upstream `now-cli`'s side *(not having to maintain two different `package.json`'s and keep them in sync as in the `mv package-homebrew.json package.json` [approve](https://github.com/Homebrew/homebrew-core/pull/14827#issuecomment-310905513)*):

We could upstream a node based build script ([like this one](https://gist.github.com/chrmoritz/c29019d322d7be4c81d6e5b2de9bafcc)), which we would [run as a first `install` step](https://gist.github.com/chrmoritz/0073afc406fc36f2b1fc46f08352d8a9/3e21ced177184c864b238558fe0bd46b09b24cf0#file-now-cli-rb-L18). It would patch the upstream provided `package.json` *(aimed for building now-cli in a node containing executable)* to a homebrew usable `package.json` *(aimed for installing it like just a normal node module with language/node with a runtime dep on our homebrew provided node 8)*. 

*Because it's a simple node script, it can parse the `package.json` and operate on a normal JavaScript Object to do much more clever things (while separating the devDeps in for our use case real deps and real devDeps) than you could do with a patch or inreplace.*

You can try it out with
```
brew install https://gist.githubusercontent.com/chrmoritz/0073afc406fc36f2b1fc46f08352d8a9/raw/3e21ced177184c864b238558fe0bd46b09b24cf0/now-cli.rb
```

What do you think about it? @ilovezfs @leo

**Comment by ilovezfs** [comment]
@chrmoritz yes that makes sense to me assuming the script is upstreamed as you said.

**Comment by amio** [comment]
> and please let us know if something else than `now help` is broken with it.

@chrmoritz That works fine in every aspect, thanks! While it seems executing source code in node instead of running a prebuilt binary, to me this feels like "an npm in brew"(download javascript and execute it using nodejs).

**Comment by ilovezfs** [comment]
@amio I suggest you spend some time reading Homebrew's documentation, and in particular https://github.com/Homebrew/brew/blob/master/docs/Acceptable-Formulae.md#we-dont-like-binary-formulae.

**Comment by amio** [comment]
Yes, I mean, we might need to try build this binary with `pkg` on homebrew's CI.

**Comment by ilovezfs** [comment]
@amio I see. Thanks for clarifying. Nonetheless, the preferred approach is to use Homebrew's node at run-time.

**Comment by ilovezfs** [comment]
Any news here?

**Comment by amio** [comment]
I've tried but getting no better results, I think we could close this for a while, until a better solution comes up.

Thank you all for helping with this!

**Comment by rauchg** [comment]
@ilovezfs would you accept a formula that runs the build process we run for `now-cli`?

Here are the reasons we use `pkg`:
- We use node as a built-in runtime and everything is still reproducible and open source
- We like the idea of having full control over what version runs. At scale, you start running into many issues when the Node.js version is not the one you tested. New Node.js versions might break things in weird ways, introduce memory leaks, etc. It's like using Node.js as a dependency, just like any other library dependency
- The zipped binary is 16mb and will likely get much smaller over time. 
- The startup time as of the latest pkg release is already 10% faster than regular Node.js, and it will likely *get much faster* later. We can pre-compile V8 snapshots, avoid disk I/O and much more.

Seems like we're picking on an implementation detail. Ultimately, now-cli is open-source, built on open-source libraries, with a reproducible build system. The output of that build is a set of binaries, like most other programs.

**Comment by ilovezfs** [comment]
@rauchg it's not totally out of the question but

>At scale, you start running into many issues when the Node.js version is not the one you tested. New Node.js versions might break things in weird ways, introduce memory leaks, etc. It's like using Node.js as a dependency, just like any other library dependency

is not sufficient justification. The same could be said of all sorts of dependencies, and we're not about to start building separate copies of everything for each package in Homebrew to speculatively prevent bugs that don't exist yet.

**Comment by chrmoritz** [comment]
> We like the idea of having full control over what version runs

You could select a major node version for the `now-cli` formula by depending on our versioned node formulae. The likelihood of breakage withing the same major node version should be reasonable small and homebrew users would jsut receive node updates (including security updates) a little bit earlier. 

> The zipped binary is 16mb and will likely get much smaller over time.

The concerns are more about install size (58mb) and less about download size. If every node module would start bundling their own node version into the Cellar we would quickly end up with installing 10+ duplicate node binaries for some users. 
*(I think this discussion is similar to wether you should statically link a (big) library into every formula requiring it or wether each of these formulae should depend on a shared version of said library and dynamically link to it.)*

> a formula that runs the build process we run for `now-cli`

Such a formula would look like this: https://gist.github.com/chrmoritz/f392d8ac8c66e008ddb660ae9ebb4065
```
brew install https://gist.githubusercontent.com/chrmoritz/f392d8ac8c66e008ddb660ae9ebb4065/raw/b423283677c1a1a73655f6f2d92916468f1e8559/now-cli.rb
```

**Comment by ilovezfs** [comment]
>You could select a major node version for the now-cli formula by depending on our versioned node formulae.

This would become a possibility _if_ it broke with node 9 in a not-easily-fixable way.

**Pull request closed without merging**
